lib/ansible/modules/source_control/gitlab_service.py

```diff
@@ -0,0 +1,400 @@
+
+
+def init_definitions():
+    for service_slug, params in RAW_SERVICES_DEFINITIONS.items():
+        for name, definition in params.items():
+            if 'type' in definition:
+                definition['type'] = bool if definition['type'] == 'bool' else int if definition['type'] == 'int' else definition['type']
+            if name in GitLabServices.credentialParams:
+                definition['no_log'] = True
+    return RAW_SERVICES_DEFINITIONS
+
+
+class GitLabServices(object):
+    HOOK_EVENTS = ['push', 'issues', 'confidential_issues', 'merge_requests', 'tag_push', 'note', 'confidential_note', 'job', 'pipeline', 'wiki_page']
+    credentialParams = ['password', 'token', 'api_key']
+
+    def __init__(self, module, git, name):
+        self._module = module
+        self.name = name
+
+    # "create" can only happen once for this service during the life of the project)
+    # merge new attributes in the object retrieved from the server
+    def create(self, remote_service, active, params, events):
+        local_service = self.asApiObject(active, params, events)
+        for k, v in local_service.items():
+            setattr(remote_service, k, v)
+        if not self._module.check_mode:
+            remote_service.save()
+        return {'before': {}, 'after': str(remote_service.attributes)}
+
+    def update(self, remote, active, params, events):
+        diff = not self.equals(remote.attributes, active, params, events)
+        if diff:
+            # remote.active = active
+            for k, v in self.expandEvents(events).items():
+                setattr(remote, k, v)
+            # when saving, params does directly in the request
+            # (remember they get retrieved under the "property" key of the remote dict())
+            for k, v in params.items():
+                setattr(remote, k, v)
+            if not self._module.check_mode:
+                remote.save()
+            return remote
+        else:
+            return False
+
+    def asApiObject(self, active, params, events):
+        obj = {'active': active}
+        obj.update(self.expandEvents(events))
+        obj.update(params)
+        return obj
+
+    def expandEvents(self, events):
+        try:
+            # See https://gitlab.com/gitlab-org/gitlab-ce/issues/58321 for why it's useful to
+            # discard unsupported events before comparing
+            supported_events = RAW_SERVICES_DEFINITIONS[self.name]['_events']
+        except KeyError:
+            supported_events = None
+        return {value + '_events': bool(value in events) for value in self.HOOK_EVENTS if supported_events is None or value in supported_events}
```

Comment:
Instead of a dict comprehension, which our CI has flagged as invalid, please use a `dict()` constructor with a generator expression:

```
dict((k, v) for k, v in thing.items())
```

Reply:
CI has flagged the Python 2.6 incompatible. As discussed before, this module depends upon Python 2.7+ for the reason that the underlying python-gitlab library is 2.7+
As such there does not seem that the >= 2.6 requirement for the PR itself applies here since the python-2.6 CI  green-light would just be misleading anyway.